dev/com.ibm.ws.transport.http/src/com/ibm/ws/http/dispatcher/internal/channel/HttpDispatcherLink.java

```diff
@@ -193,24 +194,43 @@ public void close(VirtualConnection conn, Exception e) {
         // so we will have to use close API from SRTConnectionContext31 and call closeStreams.
         String closeNonUpgraded = (String) (this.vc.getStateMap().get(TransportConstants.CLOSE_NON_UPGRADED_STREAMS));
         if (closeNonUpgraded != null && closeNonUpgraded.equalsIgnoreCase("true")) {
-
             if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {
-                Tr.debug(tc, "close streams from HttpDispatcherLink.close");
+                Tr.debug(tc, "close, CLOSE_NON_UPGRADED_STREAMS");
+            }
+
+            // Save the remain upgrading and unread data into a VC's stateMap which will be consumed in the UpgradeInputByteBufferUtil.initialRead
+            if (this.isc.isReadDataAvailable()) {
+                WsByteBuffer currentBuffer = this.isc.getReadBuffer();
```

Comment:
Isaac, this is what I have asked you about whether there is any action after `this.isc.getReadBuffer()` ? main concern is about memory leak.

Reply:
Discussed and saw that the buffer is duplicated now and the original buffer in the service context is already cleaned [here](https://github.com/OpenLiberty/open-liberty/blob/532731c5121ed3e8b43df995bb5899c7a14b7711/dev/com.ibm.ws.transport.http/src/com/ibm/ws/http/dispatcher/internal/channel/HttpDispatcherLink.java#L1200-L1216). We should be okay since I also saw the logic for cleaning up the buffer in the upgrade path seems to also affect this so I think this should be okay.